compiler/hie-core/src/Development/IDE/Core/Shake.hs

```diff
@@ -395,9 +395,13 @@ updateFileDiagnostics fp k ShakeExtras{diagnostics, state, debouncer, eventer} c
             let newDiags = getFileDiagnostics fp newDiagsStore
             pure (newDiagsStore, (newDiags, oldDiags))
         let uri = fromNormalizedUri $ filePathToUri' fp
-        when (newDiags /= oldDiags) $ do
-            let delay = if null newDiags then 0.1 else 0
-            registerEvent debouncer delay uri $ eventer $ publishDiagnosticsNotification uri newDiags
+        -- Note that we need to emit this unconditionally event if the
+        -- set of diagnostics has not changed since the last event
+        -- that had this set of diagnostics might have been killed by the
+        -- debouncer. Given that the debouncer effectively functions
+        -- as a rate limit this is fine.
```

Comment:
Yes, it rate limits them, but I suspect we are sending a notification for every file on every change, limited to 0.1 per second. We could still do duplicate suppressing, if we move the check after the debouncer. That said, I assume VS Code does duplicate suppression on its side, and can deal with huge number of events, so it might not be worth it.

Reply:
Yeah for now, I don’t think it’s worth it. I’ve added a comment on how we could reintroduce it.

Reply:
Since it ended up breaking other assumptions made by some test suites, I’ve reintroduced deduplication.